src/com/xilinx/rapidwright/rwroute/RoutableNode.java

```diff
@@ -0,0 +1,570 @@
+		this.index = index;
+		this.type = type;
+		this.node = node;
+		this.children = null;
+		this.target = false;
+		this.setEndTileXYCoordinates();
+		this.setBaseCost();
+		this.presentCongesCost = 1;
+    	this.historicalCongesCost = 1;
+    	this.setVisited(false);
+		this.usersConnectionCounts = null;
+		this.driversCounts = null;
+		this.prev = null;
+		if(timingDriven){
+			this.setDelay(RouterHelper.computeNodeDelay(delayEstimator, node));
+		}
+	}
+	
+	public int setChildren(int globalIndex, Map<Node, Routable> createdRoutable, Set<Node> reserved, RouteThruHelper routethruHelper){
+		this.children = new ArrayList<>();
+		List<Node> allDownHillNodes = this.node.getAllDownhillNodes();
+		
+		for(Node node:allDownHillNodes){		
+			if(reserved.contains(node)) continue;		
+			if(isExcluded(node, timingDriven)) continue;
+			if(routethruHelper.isRouteThru(this.node, node)) continue;
+			
+			Routable child = createdRoutable.get(node);
+			if(child == null) {
+				RoutableType type = RoutableType.WIRE;		
+				child = new RoutableNode(globalIndex++, node, type);
+				createdRoutable.put(node, child);
+			}
+			this.children.add(child);//the sink rnode of a target connection has been created up-front
+		}
+		return globalIndex;
+	}
+	
+	public void setBaseCost(){
+		if(this.type == RoutableType.WIRE){
+			baseCost = 0.4f;
+			// NOTE: IntentCode is device-dependent
+			IntentCode ic = node.getIntentCode();
+			switch(ic) {
+			case NODE_PINBOUNCE:
+				type = RoutableType.PINBOUNCE;
+				break;
+			case NODE_PINFEED:
+				type = RoutableType.PINFEED_I;
+				break;
+			
+			case NODE_DOUBLE:
+				if(this.endTileXCoordinate != this.getNode().getTile().getTileXCoordinate()) {
+					baseCost = 0.4f*this.length;
+				}
+				break;
+			case NODE_HQUAD:
+				baseCost = 0.35f*this.length;
+				break;
+			case NODE_VQUAD:
+				baseCost = 0.15f*this.length;// VQUADs have length 4 and 5
+				break;
+			case NODE_HLONG:
+				baseCost = 0.15f*this.length;// HLONGs have length 6 and 7
+				break;
+			case NODE_VLONG:
+				baseCost = 0.7f;
+				break;	
+			default:
+				if(this.length != 0) baseCost *= this.length;
+				type = RoutableType.WIRE;
+				break;
+			}	
+		}else if(this.type == RoutableType.PINFEED_I){
+			baseCost = 0.4f;
+		}else if(this.type == RoutableType.PINFEED_O){
+			baseCost = 1f;
+		}
+	}
+
+	@Override
+	public boolean overUsed() {
+		return Routable.capacity < this.getOccupancy();
+	}
+	
+	@Override
+	public boolean used(){
+		return this.getOccupancy() > 0;
+	}
+	
+	@Override
+	public boolean hasMultiDrivers(){
+		return Routable.capacity < this.numUniqueDrivers();
+	}
+
+	@Override
+	public void setEndTileXYCoordinates() {
+		Wire[] wires = this.node.getAllWiresInNode();
+		List<Tile> intTiles = new ArrayList<>();
+		for(Wire w : wires) {
+			if(w.getTile().getTileTypeEnum() == TileTypeEnum.INT) {
+				intTiles.add(w.getTile());
+			}
+		}
+		Tile endTile = null;
+		if(intTiles.size() > 1) {
+			endTile = intTiles.get(1);
```

Comment:
Is it guaranteed that the end tile is the second one in this list?

Reply:
Right, the second one in the list is the end tile. One thing should be noted is that the end X and Y coordinates could be inaccurate for some nodes that have two exit points.

Reply:
I've found one case where the logic in this function seems to fail. My sink node is `INT_X6Y84/BOUNCE_E_15_FT0` which has the following wires: `INT_X6Y84/BOUNCE_E_15_FT0, INT_X6Y85/BOUNCE_E_BLN_15_FT1, CLEL_R_X6Y84/CLE_CLE_L_SITE_0_H_I]`
The logic uses the second of the two `INT` as the end tile, when actually the sink is at `CLEL_R_X6Y84`.

Reply:
We noticed the case and it is true that the end tile found by the code can be incorrect for nodes that have two exit points. But it would not hurt the bounding box computing or the routing. It is because [every output / input pin has been mapped to its connecting INT Tile](https://github.com/YunxZhou/RapidWright/blob/a5cdce2c2d651f57c378e051e8534596e93a0417/src/com/xilinx/rapidwright/rwroute/RWRoute.java#L527), considering that each pin connects to a unique node. The mapping provides a regular coordinate system of routing resources of the architecture (using coordinates of INT tiles only). This means the CLEL_R_X6Y84 tile coordinates would not matter when computing the connection bounding box.